### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by viaviaj**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 1]
```diff
@@ -93,6 +96,7 @@ public User update(User user) {
             statement.setString(2, user.getLogin());
             statement.setString(3, user.getPassword());
             statement.setLong(4, user.getId());
+            statement.setBytes(5, user.getSalt());
```
Comment (julia-shvabovich) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: you should change 4 to 5 in setLong() and 5 to 4 in setBytes()

File: src/main/resources/init_db.sql [hunk 2]
```diff
@@ -17,11 +17,12 @@ CREATE TABLE `internet_shop`.`users`
 (
     `user_id`   BIGINT(11)   NOT NULL AUTO_INCREMENT,
     `user_name` VARCHAR(225) NOT NULL,
-    `login`     VARCHAR(225) NOT NULL,
+    user_login    VARCHAR(225) NOT NULL,
```
Comment (julia-shvabovich) [inline src/main/resources/init_db.sql]:
```suggestion
    `user_login`  VARCHAR(225) NOT NULL,
```

File: src/main/resources/init_db.sql [hunk 3]
```diff
@@ -17,11 +17,12 @@ CREATE TABLE `internet_shop`.`users`
...
     `password`  VARCHAR(225) NOT NULL,
     `deleted`   TINYINT(1)   NOT NULL DEFAULT 0,
+    `user_salt` VARBINARY(512) NOT NULL,
     PRIMARY KEY (`user_id`),
-    UNIQUE INDEX `login_UNIQUE` (`login` ASC) VISIBLE
+    UNIQUE INDEX `login_UNIQUE` (user_login ASC) VISIBLE
```
Comment (julia-shvabovich) [inline src/main/resources/init_db.sql]:
```suggestion
    UNIQUE INDEX `login_UNIQUE` (`user_login` ASC) VISIBLE
```

File: pom.xml [hunk 4]
```diff
@@ -37,6 +37,11 @@
             <artifactId>jstl</artifactId>
             <version>1.2</version>
         </dependency>
+        <dependency>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>1.2.17</version>
+        </dependency>
```
Comment (julia-shvabovich) [inline pom.xml]: Why do you need this if you don't use logger?

File: src/main/java/com/internet/shop/controllers/InjectDataController.java [hunk 5]
```diff
@@ -26,9 +26,18 @@
...
+        userService.create(user);
+
+        User admin = new User("user");
+        user.setLogin("admin");
+        user.setPassword("123");
+        user.setRoles(Set.of(Role.of("ADMIN")));
+        userService.create(user);
+
         shoppingCartService.create(new ShoppingCart(1L));
         shoppingCartService.create(new ShoppingCart(2L));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/InjectDataController.java]: you should not create a shopping cart for admin

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 6]
```diff
@@ -185,9 +189,10 @@ private void deleteRoles(Long userId) {
     private User getFromResultSet(ResultSet resultSet) throws SQLException {
         long id = resultSet.getLong("user_id");
         String name = resultSet.getString("user_name");
-        String login = resultSet.getString("login");
+        String login = resultSet.getString("user_login");
         String password = resultSet.getString("password");
-        User user = new User(name, login, password);
+        byte[] salt = resultSet.getBytes("user_salt");
+        User user = new User(name, login, password, getRoleSetByUserId(id), salt);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java]: we should not set roles here - we are opening one connection inside of another which is bad practice

File: src/main/java/com/internet/shop/security/AuthenticationServiceServiceImpl.java [hunk 7]
```diff
@@ -16,10 +17,11 @@
     public User login(String login, String password)
             throws AuthenticationException {
         Optional<User> userFromDB = userService.findByLogin(login);
-        if (userFromDB.isEmpty()
-                || !userFromDB.get().getPassword().equals(password)) {
-            throw new AuthenticationException("Incorrect login or password!");
+        if (userFromDB.isPresent()
+                && (userFromDB.get().getPassword()
+                .equals(HashUtil.hashPassword(password, userFromDB.get().getSalt())))) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/security/AuthenticationServiceServiceImpl.java]:
```suggestion
                && passwordIsValid(userFromDB.get(), password)) {
```
let's create a private method here

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -15,6 +16,10 @@
 
     @Override
     public User create(User user) {
+        byte[] salt = HashUtil.getSalt();
+        String hashedPassword = HashUtil.hashPassword(user.getPassword(),salt);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]:
```suggestion
        String hashedPassword = HashUtil.hashPassword(user.getPassword(), salt);
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/model/User.java [hunk 9]
```diff
@@ -27,6 +27,9 @@ public User(String name, String login, String password, Set<Role> roles, byte[]
     public User(String user) {
     }
 
+    public User(String name, String login, String password, byte[] salt) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/User.java:30]: there are too many constructors in this class
do you use them all?
let's maybe leave the most relevant ones and set the fields if necessary

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**